doc: update plugin spec

FeynmanZhou opened the pull request:
Update notation plugin spec based on #549 and [PRD](https://hackmd.io/1nQa69DeROyqrO_IB5yGZQ?view#Notation-plugin-management-PRD)

Added new commands and examples including:

- notation plugin install     Install a plugin
- notation plugin ninstall   Uninstall a plugin

Updated the output of `notation plugin list`.

Two-Hearts commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +42,87 @@ Aliases:
 
 Usage:
-  notation plugin install [flags] <plugin path>
+  notation plugin install [flags] <plugin location>
```
how about:
```suggestion
  notation plugin install [flags] <plugin source>
```

Two-Hearts commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +42,87 @@ Aliases:
   -h, --help                    help for install
   -f, --force                   force the installation of a plugin
-
+      --checksum                verify the checksum digest of a plugin
```
To be more precise:
```suggestion
      --checksum string               must match SHA256 of the plugin source
```

Two-Hearts commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +42,87 @@ Aliases:
-### Install a plugin
+## Install a plugin 
+
+### Install a plugin from a local file
```
```suggestion
### Install a plugin from file system
```

Two-Hearts commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +42,87 @@ Aliases:
   -f, --force                   force the installation of a plugin
-
+      --checksum                verify the checksum digest of a plugin
+      --source string           the location of plugin installation file, options: "file", "url","registry" (default "file")                  
```
To align with other designs, can we use separate flags? Something like:
```suggestion
      --file bool           if set, install the plugin from file system
      --url bool            if set, install the plugin from URL
```
We could simply make these flags mutual exclusive. (Based on PRD, we will support two flags in the first iteration.)

FeynmanZhou replied:
Having a `--source` flag with different options"file", "url","registry" (default "file") seems simpler than providing multiple flags here.

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,87 @@ Aliases:
+### Install a plugin from a local file
+
+```shell
+$ notation plugin install <path> --checksum <digest>
````
```suggestion
$ notation plugin install --file --checksum <digest> <file_path>
```

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,87 @@ Aliases:
+$ notation plugin install <path> --checksum <digest>
+```
+
+Upon successful execution, the plugin is copied to plugin directory. The name and version of the installed plugin is displayed as follows. 
````
```suggestion
Upon successful execution, the plugin is copied to Notation's plugin directory. The name and version of the installed plugin is displayed as follows. 
```

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,87 @@ Aliases:
+Successfully installed plugin <plugin name>, version <x.y.z>
+```
+
+If the plugin directory does not exist, it will be created. When an existing plugin is detected and the version is the same as the installing plugin, it fails to install and returns the error as follows. Users can use a flag `--force` to skip version check and force the installation a specified version.
````
Since we have the `notation plugin upgrade` command, we are not checking versions in the `notation plugin install` command anymore. Instead, if there is an existing plugin, the install would fail.
```suggestion
If the plugin directory does not exist, it will be created. When an existing plugin is detected, it fails to install and returns the error as follows. Users can use a flag `--force` to skip existence check and force the installation with a specified version.
```

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,87 @@ Aliases:
 
 ```shell
-notation plugin install <plugin path>
+$ notation plugin install --source url <path> --checksum <digest>
````
```suggestion
$ notation plugin install --url --checksum <digest> <URL>
```

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,87 @@ Aliases:
 ```
 
-Upon successful execution, the plugin is copied to plugins directory and name+version of plugin is displayed. If the plugin directory does not exist, it will be created. When an existing plugin is detected, the versions are compared and if the existing plugin is a lower version then it is replaced by the newer version.
+### Install a plugin as an OCI artifact from a registry
+
+```shell
+$ notation plugin install --source registry <artifact reference>
+```
````
Based on PRD, this one is planned in `v1.2.0` right?
```suggestion
```

FeynmanZhou replied:
Yes. I marked it as "for future iteration"

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,87 @@ Aliases:
+### notation plugin uninstall
 
 ```text
-Removes a plugin
+Uninstall a plugin
 
 Usage:
-  notation plugin remove [flags] <plugin name>
+  notation plugin uninstall [flags] <plugin name>
 
 Flags:
   -h, --help          help for remove
 
 Aliases:
-  remove, rm, uninstall, delete
+  remove, rm, uninstall
````
The `notation plugin uninstall` command needs to be script friendly as well. Thus, we still need a `--yes` flag to `not prompt for confirmation`.

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,87 @@ Aliases:
 
 ```shell
-notation plugin remove <plugin name>
+notation plugin uninstall <plugin name>
````
Need to add an example of:
`notation plugin uninstall <plugin name> --yes`

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,145 @@ Aliases:
 
 ```shell
-notation plugin install <plugin path>
+$ notation plugin install --source file <file_path> --checksum <digest>
````
Although it's the same, this might look better:
```suggestion
$ notation plugin install --source file --checksum <digest> <file_path>
```
Same applies to other examples.

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,145 @@ Aliases:
+Successfully installed plugin <plugin name>, version <x.y.z>
+```
+
+If the plugin directory does not exist, it will be created. When an existing plugin is detected and the version is the same as the installing plugin, it fails to install and returns the error as follows. Users can use a flag `--force` to skip existence check and force the installation with a specified version.
````
Only existence is checked, version is not checked here.
```suggestion
If the plugin directory does not exist, it will be created. When an existing plugin is detected, it fails to install and returns the error as follows. Users can use a flag `--force` to skip existence check and force the installation.
```

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,145 @@ Aliases:
+Install a Notation plugin from a registry. Users can verify the plugin's signature with `notation verify` before the plugin installation.
+
+```shell
+$ notation plugin install --source registry <artifact_reference>
````
Checksum check is always required:
```suggestion
$ notation plugin install --source registry --checksum <digest> <artifact_reference>
```

FeynmanZhou replied:
Why an OCI image requires checksum? Maybe verify the signature and its image digest are enough to ensure the plugin's integrity and authenticity.

shizhMSFT replied:
That's why an example is required.

The `artifact_reference` contains the digest.

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,145 @@ Aliases:
+$ notation plugin install --source registry <artifact_reference>
+```
+
+### Upgrade a plugin to a higher version from file system 
````
Need an example of using `notation plugin upgrade`, but the plugin does NOT exist. In this case, we should return an error.

yizha1 commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +42,145 @@ Aliases:
   -f, --force                   force the installation of a plugin
-
+      --checksum string         must match SHA256 of the plugin source
+      --source string           the location of plugin installation file, options: "file", "url","registry" (default "file")                  
```
It is possible that the type of source can be detected automatically, so that we don't need specify the option "file", "url", "registry". It is not a common practice.
- filepath
- http URL
- repo URL

A [guideline ](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md#argument-naming-conventions) for your reference

FeynmanZhou replied:
I was hoping to omit the plugin source options either and make the source type detect by Notation automatically.  @Two-Hearts mentioned there is a special case that if a file path is as same as a repo URL, it will be hard for Notation to distinguish that case.

Two-Hearts replied:
> @yizha1 I was hoping to omit the plugin source options either and make the source type detect by Notation automatically. @Two-Hearts mentioned there is a special case that if a file path is as same as a repo URL, it will be hard for Notation to distinguish that case.

To correct, a file path can be as same as a valid oci artifact reference in a registry. If we are going to support oci registry in the future, then we should consider this situation in the initial design. @yizha1 @FeynmanZhou

yizha1 replied:
@FeynmanZhou @Two-Hearts Maybe we can refer to `helm install` command, which supports installation from a file/url/repo

Two-Hearts replied:
@yizha1 That's a good idea, from helm, I could see:
`By OCI registries: helm install mynginx --version 1.2.3 oci://example.com/charts/nginx`
They ask user to add a prefix `oci://` to the input and they explicitly check for this prefix, if it exists, then it's by oci registry. @FeynmanZhou What do you think?

Two-Hearts replied:
Another option might be:
1. We check the user input in the file system first. If it's a valid, existed, file path. Then we take the input as a file path and try install from there.
2. If it does not exist in the file system, then we check if it's a valid URL. If yes, install from the URL.
3. For the current iteration, if not file path nor URL, we will return an error. For future iteration, if we support oci registry, then we treat the input as an oci reference. If it's a valid reference, then we install from oci registry.

yizha1 replied:
Helm also supports installation from a repo, how does helm distinguish it is a file or a repo? For example, 

```
$ helm repo ls
NAME                    URL
harbor                  https://helm.goharbor.io

$ helm install myharbor harbor/harbor

$ ls
<no directory harbor/harbor>
```

Two-Hearts replied:
> Helm also supports installation from a repo, how does helm distinguish it is a file or a repo? For example,
> 
> ```
> $ helm repo ls
> NAME                    URL
> harbor                  https://helm.goharbor.io
> 
> $ helm install myharbor harbor/harbor
> 
> $ ls harbor/harbor
> ls: cannot access 'harbor/harbor': No such file or directory
> ```

@yizha1 That's a good question. The answer is, helm treats `harbor/harbor` as a file path first, they check if this is a valid file path. If the path does not exist, then they take it as a repo. So the process looks automatically on the external user's point of view.

Two-Hearts replied:
> @Two-Hearts If the edge case can be covered by this option, it sounds good to me.

Yes, the edge case can be covered. No matter what user input it is, Notation always check by order: file path -> URL [-> OCI reference, in the future].

yizha1 reviewed (commented):
Thanks @FeynmanZhou . I did a first round review, will take another pass later.

FeynmanZhou commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +42,145 @@ Aliases:
-  -h, --help          help for remove
+  -h, --help                    help for install
+      --checksum string         must match SHA256 of the plugin source
+      --source string           the location of plugin installation file, options: "file", "url","registry" (default "file")                  
```
@yizha1  In general, a feature spec can have features in the roadmap or future milestones. The Notary Project spec also has some points and feature requirements that are planned for future releases. 
If we have a feature defined in the spec earlier, it is possible to be implemented anytime. I think we can mark this kind of features as "for future milestone" in spec to avoid confusion.

yizha1 replied:
My point is that this section will be used for implementation, however, we won't support registry, so we cannot display it to users, right?

shizhMSFT commented on specs/commandline/plugin.md:
```diff
@@ -15,9 +15,11 @@ Usage:
-  remove      Removes a plugin
+  uninstall   Uninstall a plugin
+  upgrade     Upgrade a plugin
+  
```
remove extra line

shizhMSFT commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +43,151 @@ Aliases:
   -f, --force                   force the installation of a plugin
-
+      --checksum string         must match SHA256 of the plugin source
+      --source string           the plugin installation source, options: "file", "url","registry" (default "file")                  
 Aliases:
```
missing extra line

shizhMSFT commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +43,151 @@ Aliases:
   -f, --force                   force the installation of a plugin
-
+      --checksum string         must match SHA256 of the plugin source
+      --source string           the plugin installation source, options: "file", "url","registry" (default "file")                  
```
should we default to `registry`? or even `auto`?

shizhMSFT commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +43,151 @@ Aliases:
+  -y, --yes           do not prompt for confirmation
 Aliases:
-  remove, rm, uninstall, delete
+  remove, rm, uninstall
```
nit:
```suggestion
  uninstall, remove, rm
```

shizhMSFT commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +43,151 @@ Aliases:
+Install a Notation plugin from file system and verify the plugin checksum.
+
+```shell
+$ notation plugin install --source file --checksum <digest> <file_path>
````
When install from a local file, the checksum should be optional. Otherwise, people will just do
```bash
notation plugin install --source file --checksum $(sha256sum $file | cut -f1 -d' ') $file
```
which is not really useful.

shizhMSFT commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +43,151 @@ Aliases:
   -h, --help                    help for install
   -f, --force                   force the installation of a plugin
-
+      --checksum string         must match SHA256 of the plugin source
```
What's the format of `checksum`?

FeynmanZhou replied:
It's the encoded value without `sha256:`.

shizhMSFT commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +43,151 @@ Aliases:
 
 ```shell
-notation plugin install <plugin path>
+$ notation plugin install --source url <URL> --checksum <digest>
````
What kind of URL does it support?

shizhMSFT commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +43,151 @@ Aliases:
+If the plugin is not found, an error is returned showing the syntax for the plugin list command to show the installed plugins.
+
+```shell
+Error: <plugin_name> does not exist. Please install the plugin using 'notation plugin install' first.
````
Should we suggest `notation plugin ls` instead?

FeynmanZhou commented:
@rgnote @priteshbandi @iamsamirzon  Would you mind reviewing the notation plugin mgmt spec updates? Thanks

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,144 @@ Aliases:
+Error: failed to install the plugin, <plugin_name> already installed
+```
+
+If the entered plugin checksum digest doesn't match the published checksum, Notation will return an error message and will not start installation.
````
One thing might be worth to point out in the spec:
`when user is installing a plugin from file system, the checksum is optional.`

FeynmanZhou replied:
Agree. Added this note

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,144 @@ Aliases:
 
 ```shell
-notation plugin install <plugin path>
+$ notation plugin install --checksum <digest>
````
This should be:
```suggestion
$ notation plugin install --checksum <digest> <URL>
```

FeynmanZhou replied:
Good catch. Fixed

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,144 @@ Aliases:
+Upgrade a Notation plugin to a higher version from file system and verify the plugin checksum.
+
+```shell
+$ notation plugin upgrade <file_path> --checksum <digest>
````
When upgrading from a file system, `--checksum` should be optional? And should we also provide examples when upgrading from URL and OCI registry?

FeynmanZhou replied:
Yes. Removed checksum validation here. Added two examples for upgrading plugin from URL and OCI registry

Two-Hearts reviewed (dismissed):
LGTM

priteshbandi commented on specs/commandline/plugin.md:
```diff
@@ -99,6 +209,6 @@ An example of output from `notation plugin list`:
-azure-kv                               Sign artifacts with keys in Azure Key Vault   v0.5.0-rc.1   [SIGNATURE_GENERATOR.RAW]                                                                                <nil>
-com.amazonaws.signer.notation.plugin   AWS Signer plugin for Notation                1.0.290       [SIGNATURE_GENERATOR.ENVELOPE SIGNATURE_VERIFIER.TRUSTED_IDENTITY SIGNATURE_VERIFIER.REVOCATION_CHECK]   <nil>
+azure-kv                               Sign artifacts with keys in Azure Key Vault   v1.0.0        Signature generation                                                                                <nil>
+com.amazonaws.signer.notation.plugin   AWS Signer plugin for Notation                1.0.290       Signature envelope generation, Trusted Identity validation, Certificate chain revocation check   <nil>
```
Should we keep capabilities as enum instead of text as they are more for plugin publisher as compared to consumer?

FeynmanZhou replied:
The original enum could be used and kept on the [plugin-extensibility.md#plugin-metadata)](https://github.com/notaryproject/specifications/blob/main/specs/plugin-extensibility.md#plugin-metadata). For the output of `notation plugin list`, I think it should be user-friendly so using text would be much straightforward.

priteshbandi commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,160 @@ Aliases:
+If the plugin is not found, an error is returned showing the syntax for the plugin list command to show the installed plugins.
+
+```shell
+Error: <plugin_name> does not exist. Please check the plugin information using 'notation plugin list' first.
````
```suggestion
Error: unable to find plugin plugin_name. To view a list of installed plugins, use `notation plugin list`.
```

priteshbandi commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,160 @@ Aliases:
 
 ```shell
-notation plugin install <plugin path>
+$ notation plugin install <file_path>
````
Can we please specify the file supported file formats and install behavior for them ?
For Example:

1. `.zip`: The ZIP file should house the plugin in the root directory. If the ZIP archive contains multiple files, Notation will locate an executable file associated with the plugin and copy it, ignoring any other files?
2. `.tar.gz`: Similar for .tar.gz
3. `.tar`: Should we even support?
4. `.gz`: Should we even support?
5. All other format, Notation will treat them as executable?

^ Notation will use file extension to determine how to unpack file/archive ?

Two-Hearts replied:
@priteshbandi These questions are related to implementation though. Since you asked, here are the current design:
1. For `.zip`, Notation goes through the zip file -> find an executable file -> validate its name (must be `notatin-{plugin-name}`) -> check if plugin already exists (if --force is not set) -> extract the executable file into a `tmp` dir (regardless the final result is a success or a failure, this `tmp` dir will be removed in the end) -> validate the executable file is indeed a valid Notation plugin by calling its `get-plugin-metada` -> at this point, copy the file into correct Notation dir with all parent dir created along the way. A special note for `.zip` file, a normal file such as `LICENSE` can be treated as executable too. That's why we cannot directly use whatever executable file Notation finds. 
2. For `.tar.gz`, Notation will first detect that a `.gz` file is passed in. Then, Notation will require this `.gz` file compressed a `.tar` file in it. If yes, then apply similar logic as `.zip`.
3. `.tar` alone is not supported.
4. `.gz` alone is not supported.
5. All other format is not very different from `.tar` or `.gz`, they are not supported.

And yes, Notation has the capability of telling apart `.zip` and `.tar.gz`, but not using file extension. Instead, it considers at most the first 512 bytes of data and determine the file type.

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,160 @@ Aliases:
+Install a Notation plugin from a remote shared address and verify the plugin checksum. Notation only supports installing plugins from an HTTPS URL.
+
+```shell
+$ notation plugin install --checksum <digest> <URL>
````
@priteshbandi I could answer this one, here's the order (similar to [helm](https://github.com/helm/helm/blob/d70e293e5fea032e843c29d15c7e85e78b0aca6b/pkg/action/install.go#L718)):
1. We check the user input in the file system first. If it's a valid, existed, file path. Then we take the input as a file path and try install from there. (Notation would validate if it's indeed a valid Notation plugin before installation.)
2. If it does not exist in the file system, then we check if it's a valid URL (with prefix `https://`). If yes, try install from the URL.

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,160 @@ Aliases:
+Install a Notation plugin from a registry. Users can verify the plugin's signature with `notation verify` before the plugin installation.
+
+```shell
+$ notation plugin install <registry>/<repository>@<digest>
````
@FeynmanZhou Since OCI related logic is planned in future iteration, can we remove it from this PR for now?

FeynmanZhou replied:
I am okay with removing it from this PR. In order to reserve the design doc, I think we can add the original PRD doc to another place (e.g. notaryproject/specification) to archive our design and discussion centrally.

yizha1 replied:
@FeynmanZhou we can create an issue for the future backlog. The future backlog will be visit during planning

FeynmanZhou replied:
Removed "Install a Notation plugin from a registry"

FeynmanZhou commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +42,160 @@ Aliases:
 
 ```text
-Removes a plugin
+Upgrade a plugin
````
I also expect allow users to upgrade without specifying the URL. It seems users still need to manually get the sha256sum from the plugin download page

FeynmanZhou replied:
> How does the proposed behavior differ from using the install command?

I expect upgrade a plugin without specifying the plugin URL. This differs from the `notation plugin install` experience. I will update the CLI spec to describe this experience. 

> Could the install command be used to handle the following scenarios:

For 1, it will be handled by current design.
For 2 and 3,  the logic is a little different. `notation plugin install` only checks the plugin existence instead of version.  
the current design and behavior are described at [this line](https://github.com/notaryproject/notation/pull/809/files#diff-7637fbf27d1c8c25ef1a9806613d4aa17aaa61d99aae29f8c802b5e50bcd2efdR105). In short, when an existing plugin is detected, it fails to install and returns the error. Users can use a flag `--force` to skip existence check and force the installation.

FeynmanZhou replied:
@priteshbandi I have adopted your suggestions above and update the `notation plugin install` usage and examples accordingly. Thanks

Two-Hearts commented:
@FeynmanZhou Can you create an issue in https://github.com/notaryproject/specifications to keep track that we are also going to update spec there? (spec related to what and how should a vendor publish a plugin).  And link this PR to the issue. Thanks.

FeynmanZhou commented:
> @FeynmanZhou Can you create an issue in https://github.com/notaryproject/specifications to keep track that we are also going to update spec there? (spec related to what and how should a vendor publish a plugin). And link this PR to the issue. Thanks.

Thanks @patrickzheng200 . Issue created: https://github.com/notaryproject/specifications/issues/284

FeynmanZhou commented:
Hi @notaryproject/notaryproject-notation-maintainers 

I have major updates in a recent commit based on the community meeting discussion on Nov 13:
- Removed `notation plugin upgrade` from this iteration. This will be considered in the next iteration
- Added two flags `--url` and `file` to `notation plugin install`, added more usage examples and error messages accordingly
- Removed "Install a Notation plugin from a registry (for future iteration)". This will be considered in the next iteration

Two-Hearts commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +41,115 @@ Aliases:
 
 Usage:
-  notation plugin install [flags] <plugin path>
+  notation plugin install [flags] <plugin_source>
```
Since one and only one of [--file, --url] is required,
```suggestion
  notation plugin install [flags] <--file|--url> <plugin_source>
```

Two-Hearts commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +41,115 @@ Aliases:
+  notation plugin install [flags] <plugin_source>
 
 Flags:
+      --file                    install a plugin from a local file
```
```suggestion
      --file                    if set, install plugin from a file in file system
```

Two-Hearts commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +41,115 @@ Aliases:
   -h, --help                    help for install
   -f, --force                   force the installation of a plugin
+      --sha256sum string        must match SHA256 of the plugin source
+      --url                     install a plugin from a remote HTTPS address
```
```suggestion
      --url                     if set, install plugin from a HTTPS URL
```

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +41,115 @@ Aliases:
+If the entered plugin checksum digest doesn't match the published checksum, Notation will return an error message and will not start installation.
+
+```console
+Error: failed to install the plugin, input checksum does not match the published checksum, expected <digest>
````
```suggestion
Error: failed to install the plugin: plugin checksum does not match user input. Expecting <sha256sum>
```

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +41,115 @@ Aliases:
+If the plugin version is higher than the existing plugin, Notation will start installation and overwrite the existing plugin.
+
+```console
+Successfully installed plugin <plugin name>, updated the version from <x.y.z> to <a.b.c>
````
The success message can be the same as before,
```suggestion
Successfully installed plugin <plugin name>, version <a.b.c>
```

FeynmanZhou replied:
I think it's necessary to tell users about the version update if users install a higher version and overwrite an older version. @yizha1 @shizhMSFT what's your suggestion for this case?

yizha1 replied:
I agree with you @FeynmanZhou

Two-Hearts replied:
> I think it's necessary to tell users about the version update if users install a higher version and overwrite an older version.

SGTM

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +41,115 @@ Aliases:
+If the plugin version is equal to the existing plugin, Notation will not start installation and return the following message. Users can use a flag `--force` to skip plugin version check and force the installation.
+
+```console
+Plugin <plugin name>-<x.y.z> already exists. 
````
```suggestion
Plugin <plugin name> with version <x.y.z> already installed.
```

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +41,115 @@ Aliases:
+```console
+Plugin <plugin name>-<x.y.z> already exists. 
+To view a list of installed plugins, use "notation plugin list".
+If you want to install the same version, use "notation plugin uninstall" to uninstall it before re-installing the same version or use "--force" to force the installation
````
we don't need to explicitly say it,
```suggestion
```

yizha1 replied:
Agree with @Two-Hearts

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +41,115 @@ Aliases:
+If the plugin version is lower than the existing plugin, Notation will return an error message and will not start installation. Users can use a flag `--force` to skip plugin version check and force the installation.
+
+```console
+Error: failed to install the plugin, plugin version is lower than the existing plugin. 
````
```suggestion
Error: failed to install the plugin: <plugin-name> current version <x.y.z> is larger than the installing version <a.b.c>
```

yizha1 replied:
@patrickzheng200 larger-->higher

FeynmanZhou replied:
Updated to `Error: failed to install the plugin: <plugin-name>. The installing version <a.b.c> is lower than the existing plugin. `

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +41,115 @@ Aliases:
+```console
+Error: failed to install the plugin, plugin version is lower than the existing plugin. 
+To view a list of installed plugins, use "notation plugin list".
+If you want to install an old version, use "notation plugin uninstall" to uninstall the existing plugin before installation or use "--force" to force the installation
````
same here,
```suggestion
```

yizha1 replied:
For older version, it is worth mentioning, but it's better we highlights not recommended, but in rare case, if users want to do that.

FeynmanZhou replied:
Installing an older version is an downgrade operation. Users may consider downgrade for some reasons. I agree with @yizha1 that mentioned it but state that installing an older version is not recommended.

FeynmanZhou replied:
Updated

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +41,113 @@ Aliases:
+
+```console
+Error: failed to install the plugin: <plugin-name>. The installing version <a.b.c> is lower than the existing plugin. 
+To view a list of installed plugins, use "notation plugin list". It is not recommended to install older versions. Use "--force" to force the installation if you want to install an old version in a certain scenario.
````
This is still a bit too verbose. And logically, I think we should move `To view a list of installed plugins, use "notation plugin list".` to the end of the message. How about,
```suggestion
It is not recommended to install an older version. Use "--force" to force the installation if you want to do so. To view a list of installed plugins, use "notation plugin list".
```

Two-Hearts commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +41,113 @@ Aliases:
+  notation plugin install [flags] <--file|--url> <plugin_source>
 
 Flags:
+      --file                    install plugin from a file in file system
   -h, --help                    help for install
   -f, --force                   force the installation of a plugin
+      --sha256sum string        must match SHA256 of the plugin source
+      --url                     install plugin from an HTTPS URL
```
nit:
```suggestion
  -d, --debug              debug mode
      --file               install plugin from a file in file system
      --force              force the installation of a plugin
  -h, --help               help for install
      --sha256sum string   must match SHA256 of the plugin source
      --url                install plugin from an HTTPS URL
  -v, --verbose            verbose mode
```

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +41,113 @@ Aliases:
+If the plugin version is lower than the existing plugin, Notation will return an error message and will not start installation. Users can use a flag `--force` to skip plugin version check and force the installation.
+
+```console
+Error: failed to install the plugin: <plugin-name>. The installing version <a.b.c> is lower than the existing plugin. 
````
```suggestion
Error: failed to install the plugin: <plugin-name>. The installing version <a.b.c> is lower than the existing plugin version <x.y.z>. 
```

FeynmanZhou replied:
Updated

Two-Hearts reviewed (dismissed):
LGTM

yizha1 reviewed (approved):
LGTM but IANAM

priteshbandi commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +41,115 @@ Aliases:
+
+If the plugin version is equal to the existing plugin, Notation will not start installation and return the following message. Users can use a flag `--force` to skip plugin version check and force the installation.
+
+```console
+Plugin <plugin name> with version <x.y.z> already exists. 
+To view a list of installed plugins, use "notation plugin list"
````
Why do we want to show user that they can use plugin list command?
```suggestion
```console
Error: Plugin <plugin name> with version <x.y.z> already exists. 
```

priteshbandi commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +41,115 @@ Aliases:
+If the plugin version is lower than the existing plugin, Notation will return an error message and will not start installation. Users can use a flag `--force` to skip plugin version check and force the installation.
+
+```console
+Error: failed to install the plugin: <plugin-name>. The installing version <a.b.c> is lower than the existing plugin version <x.y.z>.
+It is not recommended to install an older version. Use "--force" to force the installation if you want to do so. To view a list of installed plugins, use "notation plugin list"
````
```suggestion
Unable to install the plugin: <plugin-name>. The version you are trying to install (<a.b.c>) is older than the version that is already installed (<x.y.z>). To force the installation, use the "--force" option.
```

FeynmanZhou replied:
@priteshbandi  To align the behavior with the case when installing version is equal to exist plugin as you described in https://github.com/notaryproject/notation/pull/809#discussion_r1396788077, I think this output should be stderr either.

priteshbandi replied:
I concur, updated messsage
```
Error: Unable to install the plugin: <plugin-name>. The version you are trying to install (<a.b.c>) is equal or lower than the version that is already installed (<x.y.z>). To force the installation, use the "--force" option.
```

priteshbandi commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +41,115 @@ Aliases:
+### Install a plugin from URL
 
-Upon successful execution, the plugin is copied to plugins directory and name+version of plugin is displayed. If the plugin directory does not exist, it will be created. When an existing plugin is detected, the versions are compared and if the existing plugin is a lower version then it is replaced by the newer version.
+Install a Notation plugin from a remote shared address and verify the plugin checksum. Notation only supports installing plugins from an HTTPS URL.
```
```suggestion
Installs a Notation plugin from a remote location and verify the plugin checksum. Notation only supports installing plugins from an HTTPS URL, which means that the URL must start with "https://".
```

priteshbandi commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +41,115 @@ Aliases:
+Install a Notation plugin from a remote shared address and verify the plugin checksum. Notation only supports installing plugins from an HTTPS URL.
+
+```shell
+$ notation plugin install --sha256sum <digest> --url <HTTPS_URL>
````
Should we keep `--sha256sum` as optional flag ?

FeynmanZhou replied:
Notation as a security project, I think check `--sha256sum` provides security guarantee to end users. 
Alternatively, users can download a plugin file to local file system and install it from file since `notation plugin install --file` doesn't require checksum.

priteshbandi replied:
Plugin venders may sign the artifact(i.e. zip) to generate a signature (either GPG or platform specific), making sha256 mandatory doesn't looks right for this usecase.

Two-Hearts commented on specs/commandline/plugin.md:
````diff
@@ -41,51 +41,114 @@ Aliases:
+If the plugin version is equal to the existing plugin, Notation will not start installation and return the following message. Users can use a flag `--force` to skip plugin version check and force the installation.
+
+```console
+Error: Plugin <plugin name> with version <x.y.z> already exists.
````
```suggestion
Error: failed to install the plugin: <plugin-name> with version <x.y.z> already exists.
```

Two-Hearts commented on specs/commandline/plugin.md:
```diff
@@ -41,51 +41,114 @@ Aliases:
 Flags:
   -h, --help          help for remove
-
+  -y, --yes           do not prompt for confirmation
```
nit:
```suggestion
  -d, --debug     debug mode
  -h, --help      help for uninstall
  -v, --verbose   verbose mode
  -y, --yes       do not prompt for confirmation
```

priteshbandi reviewed (approved):
LGTM

Two-Hearts reviewed (approved):
LGTM
